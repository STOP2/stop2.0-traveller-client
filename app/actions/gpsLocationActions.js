export const SET_GPS_LOCATION = 'SET_GPS_LOCATION'
export const GPS_LOCATION_ERROR = 'GPS_LOCATION_ERROR'
export const REQUEST_GPS_LOCATION = 'REQUEST_GPS_LOCATION'
export const CLEAR_LOCATION = 'CLEAR_LOCATION'

let watchId = false

export let setGpsLocation = function(gpsLocationData)
{
    return {
        type: SET_GPS_LOCATION,
        gettingGpsLocation: false,
        gpsLocationData: gpsLocationData
    }
}

export let requestGpsLocation = function()
{
    return {
        type: REQUEST_GPS_LOCATION,
        gettingGpsLocation: true
    }
}

export let gpsLocationError = function(error)
{
    return {
        type: GPS_LOCATION_ERROR,
        error: error,
        gettingGpsLocation: false
    }
}

export let getGpsLocation = function()
{
    return dispatch =>
   {
        if (!watchId)
        {
            dispatch(requestGpsLocation())
            navigator.geolocation.getCurrentPosition((position) =>
            {
                dispatch(setGpsLocation(position.coords))
            },
            (error) =>
            {
                dispatch(gpsLocationError(error))
            },
                {
                    enableHighAccuracy: false,
                    timeout: 60000,
                    maximumAge: 1000
                })


            watchId = navigator.geolocation.watchPosition((position) =>
            {
                dispatch(setGpsLocation(position.coords))
            },
              (error) =>
              {
              },
                {
                    enableHighAccuracy: true,
                    timeout: 60000,
                    maximumAge: 1000
                })
        }
    }
}

export let clearWatchLocation = function()
{
    navigator.geolocation.clearWatch(watchId)
    watchId = false

    return { type: CLEAR_LOCATION }
}
