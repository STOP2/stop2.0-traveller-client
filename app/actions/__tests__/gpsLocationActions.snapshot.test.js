
import mockStore from 'redux-mock-store';

import { setGpsLocation, getGpsLocation, requestGpsLocation, gpsLocationError } from '../gpsLocationActions';

const store = mockStore();

const mockLocationSuccess = function mockLocationSuccess(geoSuccess, geoError, geoOptions) {
  geoSuccess({
    coords: {
      latitude: 60.203978,
      longitude: 24.9633573,
    },
  });
};

const mockLocationError = function mockLocationError(geoSuccess, geoError, geoOptions) {
  geoError('Failed to get location');
};

beforeEach(() => {
  store.clearActions();
});

describe('gps location actions', () => {
  it('creates a SET_GPS_LOCATION action', () => {
    expect(setGpsLocation({ coords: [] })).toMatchSnapshot();
  });

  it('creates a REQUEST_GPS_LOCATION action', () => {
    expect(requestGpsLocation()).toMatchSnapshot();
  });

  it('creates a GPS_LOCATION_ERROR action', () => {
    expect(gpsLocationError(true)).toMatchSnapshot();
  });

  it('returns a location object', async () => {
    await store.dispatch(getGpsLocation(mockLocationSuccess));

    expect(store.getActions()).toMatchSnapshot();
  });

  it('returns a location object', async () => {
    await store.dispatch(getGpsLocation(mockLocationError));

    expect(store.getActions()).toMatchSnapshot();
  });
});
