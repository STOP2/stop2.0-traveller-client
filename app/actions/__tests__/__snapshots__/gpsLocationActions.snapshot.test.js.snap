exports[`gps location actions creates a GPS_LOCATION_ERROR action 1`] = `
Object {
  "error": true,
  "gettingGpsLocation": false,
  "type": "GPS_LOCATION_ERROR",
}
`;

exports[`gps location actions creates a REQUEST_GPS_LOCATION action 1`] = `
Object {
  "gettingGpsLocation": true,
  "type": "REQUEST_GPS_LOCATION",
}
`;

exports[`gps location actions creates a SET_GPS_LOCATION action 1`] = `
Object {
  "gettingGpsLocation": false,
  "gpsLocationData": Object {
    "coords": Array [],
  },
  "type": "SET_GPS_LOCATION",
}
`;
